core/src/mast/serialization/basic_block_data_builder.rs

```diff
@@ -0,0 +1,291 @@
+                | AdviceInjector::SmtPeek
+                | AdviceInjector::U32Clz
+                | AdviceInjector::U32Ctz
+                | AdviceInjector::U32Clo
+                | AdviceInjector::U32Cto
+                | AdviceInjector::ILog2
+                | AdviceInjector::MemToMap
+                | AdviceInjector::HpermToMap => (),
+            },
+            Decorator::AsmOp(assembly_op) => {
+                self.data.push(assembly_op.num_cycles());
+                self.data.write_bool(assembly_op.should_break());
+
+                // context name
+                {
+                    let str_index_in_table =
+                        self.string_table_builder.add_string(assembly_op.context_name());
+                    self.data.write_usize(str_index_in_table);
+                }
+
+                // op
+                {
+                    let str_index_in_table = self.string_table_builder.add_string(assembly_op.op());
+                    self.data.write_usize(str_index_in_table);
+                }
+            }
+            Decorator::Debug(debug_options) => match debug_options {
+                DebugOptions::StackTop(value) => self.data.push(*value),
+                DebugOptions::MemInterval(start, end) => {
+                    self.data.extend(start.to_le_bytes());
+                    self.data.extend(end.to_le_bytes());
+                }
+                DebugOptions::LocalInterval(start, second, end) => {
+                    self.data.extend(start.to_le_bytes());
+                    self.data.extend(second.to_le_bytes());
+                    self.data.extend(end.to_le_bytes());
+                }
+                DebugOptions::StackAll | DebugOptions::MemAll => (),
+            },
+            Decorator::Event(value) | Decorator::Trace(value) => {
+                self.data.extend(value.to_le_bytes())
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct StringTableBuilder {
+    table: Vec<StringRef>,
+    str_to_index: BTreeMap<RpoDigest, StringIndex>,
+    strings_data: Vec<u8>,
+}
+
+impl StringTableBuilder {
+    pub fn add_string(&mut self, string: &str) -> StringIndex {
+        if let Some(str_idx) = self.str_to_index.get(&Rpo256::hash(string.as_bytes())) {
+            // return already interned string
+            *str_idx
+        } else {
+            // add new string to table
+            // NOTE: these string refs' offset will need to be shifted again in `into_buffer()`
+            let str_ref = StringRef {
+                offset: self
+                    .strings_data
+                    .len()
+                    .try_into()
+                    .expect("strings table larger than 2^32 bytes"),
+                len: string.len().try_into().expect("string larger than 2^32 bytes"),
+            };
+            let str_idx = self.table.len();
+
+            self.strings_data.extend(string.as_bytes());
+            self.table.push(str_ref);
```

Comment:
The strings that we encode have a lot of repetition, (i.e. `AssemblyOp.context_name` is going to be the same for a bunch of operations, and similarly for `AssemblyOp.op`). If we stored the string length in the data buffer, we'd be duplicating that length byte (assuming a small string) over and over. Given that most strings are duplicates of one another, I suspect these length bytes could add up to a nontrivial amount. Storing the length byte in `StringRef` removes that duplication.

Reply:
Maybe I'm misunderstanding how the strings table works, but don't we store only a single copy of a given string? If so, I was thinking that the length would be stored together with that string - so, there would be only one copy of the length stored as well.

Reply:
Oh no, I'm re-reading what I wrote and I think I was just confused... Applied your suggestion.

I kept the `StringRef` type, but it could be potentially cleaner to just use `DataOffset` (which makes it clear that a string reference is simply an offset into the data buffer)? I'm happy with both approaches.

Reply:
Yeah - I think using `DataOffset` is probably a bit cleaner.

Reply:
Switched to `DataOffset`